File: solutions/java/concurrency/OddEven.java
public class OddEven {

  //@include
  static class OddEvenMonitor {
    public static final boolean ODD_TURN = true;
    public static final boolean EVEN_TURN = false;
    private boolean turn = ODD_TURN;

    public synchronized void waitTurn(boolean oldTurn) {
      while (turn != oldTurn) {
        try {
          wait();
        } catch(Exception e) {
        }
      }
    }

    public synchronized void toggleTurn() {
      turn ^= true;
      notify();
    }
  }

  static class OddThread extends Thread {
    private final OddEvenMonitor monitor;

    public OddThread(OddEvenMonitor monitor) {
      this.monitor = monitor;
    }
    @Override
      public void run() {
        for (int i = 1; i <= 100; i+=2) {
          monitor.waitTurn(OddEvenMonitor.ODD_TURN);
          System.out.println(i);
          monitor.toggleTurn();
        }
      }
  }

  static class EvenThread extends Thread {
    private final OddEvenMonitor monitor;

    public EvenThread(OddEvenMonitor monitor) {
      this.monitor = monitor;
    }
    @Override
      public void run() {
        for (int i = 2; i <= 100; i+=2) {
          monitor.waitTurn(OddEvenMonitor.EVEN_TURN);
          System.out.println(i);
          monitor.toggleTurn();
        }
      }
  }
  //@exclude

  public static void main (String[] args) {
    OddEvenMonitor monitor = new OddEvenMonitor();
    Thread t1 = new OddThread(monitor);
    Thread t2 = new EvenThread(monitor);
    t1.start();
    t2.start();
    try {
      t1.join();
      t2.join();
    } catch(Exception e) {
    }
  } 
}

File: static/python/odd_even.py
import threading


# @include
class OddEvenMonitor(threading.Condition):

    ODD_TURN = True
    EVEN_TURN = False

    def __init__(self):
        super().__init__()
        self.turn = self.ODD_TURN

    def wait_turn(self, old_turn):
        with self:
            while self.turn != old_turn:
                self.wait()

    def toggle_turn(self):
        with self:
            self.turn ^= True
            self.notify()


class OddThread(threading.Thread):

    def __init__(self, monitor):
        super().__init__()
        self.monitor = monitor

    def run(self):
        for i in range(1, 101, 2):
            self.monitor.wait_turn(OddEvenMonitor.ODD_TURN)
            print(i)
            self.monitor.toggle_turn()


class EvenThread(threading.Thread):

    def __init__(self, monitor):
        super().__init__()
        self.monitor = monitor

    def run(self):
        for i in range(2, 101, 2):
            self.monitor.wait_turn(OddEvenMonitor.EVEN_TURN)
            print(i)
            self.monitor.toggle_turn()
# @exclude


def main():
    monitor = OddEvenMonitor()
    t1 = OddThread(monitor)
    t2 = EvenThread(monitor)
    t1.start()
    t2.start()
    t1.join()
    t2.join()


if __name__ == '__main__':
    main()
